File: scripts/clawtributor-campaign50/lanes/blank-numeric/verify-source.py
import hashlib
import json
import pathlib
import subprocess
import sys

lane = pathlib.Path(sys.argv[1])
phase = sys.argv[2]
assert phase in ("clean", "baseline", "candidate")
manifest = json.loads((lane / "MANIFEST.json").read_text())
git = lambda *args: subprocess.check_output(["git", *args], text=True).strip()
assert git("rev-parse", "HEAD") == manifest["base"]
assert subprocess.check_output(["node", "--version"], text=True).strip() == "v" + manifest["node"]
assert subprocess.check_output(["pnpm", "--version"], text=True).strip() == manifest["pnpm"]
assert json.loads(pathlib.Path("package.json").read_text())["packageManager"] == manifest["packageManager"]
assert json.loads(pathlib.Path("package.json").read_text())["bin"]["openclaw"] == manifest["entry"]
assert hashlib.sha256(pathlib.Path(manifest["entry"]).read_bytes()).hexdigest() == manifest["entry_sha256"]
if phase == "clean":
    assert not git("status", "--porcelain", "--untracked-files=normal")
else:
    for name, expected in manifest[phase + "_files"].items():
        assert hashlib.sha256(pathlib.Path(name).read_bytes()).hexdigest() == expected, name
    actual = set(git("diff", "--name-only", "HEAD").splitlines())
    expected = set(manifest["candidate_files"])
    if phase == "baseline":
        expected = {name for name in expected if name.endswith(".test.ts")}
    assert actual == expected, actual ^ expected
    assert not git("ls-files", "--others", "--exclude-standard")
print(json.dumps({"phase": phase, "base": manifest["base"], "node": manifest["node"], "pnpm": manifest["pnpm"], "source_hashes": "verified"}))

File: scripts/clawtributor-campaign50/lanes/blank-status-logs-green/verify-source.mjs
import assert from "node:assert/strict";
import { execFileSync } from "node:child_process";
import { createHash } from "node:crypto";
import fs from "node:fs";
import path from "node:path";
const [target, lane, phase] = process.argv.slice(2);
assert(["clean", "baseline", "candidate"].includes(phase));
const packet = JSON.parse(fs.readFileSync(path.join(lane, "PACKET.json"), "utf8"));
const hash = (bytes) => createHash("sha256").update(bytes).digest("hex");
const git = (...args) => execFileSync("git", args, { cwd: target, encoding: "utf8" }).trim();
assert.equal(process.version, packet.node);
assert.equal(process.platform, "linux");
assert.equal(git("rev-parse", "HEAD"), packet.source);
const overlay =
  phase === "candidate" ? packet.candidateHashes : phase === "baseline" ? packet.testHashes : {};
assert.equal(git("ls-files", "--others", "--exclude-standard"), "");
assert.deepEqual(
  git("diff", "--name-only", "HEAD").split("\n").filter(Boolean).sort(),
  Object.keys(overlay).sort(),
);
assert.equal(
  JSON.parse(fs.readFileSync(path.join(target, "package.json"), "utf8")).packageManager,
  packet.packageManager,
);
for (const [file, expected] of Object.entries({ ...packet.sourceHashes, ...overlay })) {
  assert.equal(hash(fs.readFileSync(path.join(target, file))), expected, file);
}
for (const [file, expected] of Object.entries(packet.files)) {
  assert.equal(hash(fs.readFileSync(path.join(lane, file))), expected, file);
}
const built = [];
let inventoryComplete = false;
let inventoryFailure;
function isPluginDependencyLink(file) {
  const parts = file.split(path.sep);
  if (
    parts[0] !== "dist" ||
    parts[1] !== "extensions" ||
    parts[2] === "node_modules" ||
    parts[3] !== "node_modules"
  )
    return false;
  const name = parts[4];
  if (!name) return false;
  if (parts.length === 6) return name.startsWith("@") && Boolean(parts[5]);
  return (
    parts.length === 5 && (name === ".bin" || (!name.startsWith(".") && !name.startsWith("@")))
  );
}
function visit(dir) {
  for (const entry of fs
    .readdirSync(path.join(target, dir), { withFileTypes: true })
    .sort((a, b) => a.name.localeCompare(b.name))) {
    const file = path.join(dir, entry.name);
    if (entry.isDirectory()) visit(file);
    else if (entry.isFile())
      built.push([file, "file", hash(fs.readFileSync(path.join(target, file)))]);
    else if (entry.isSymbolicLink()) {
      assert(isPluginDependencyLink(file), `Unexpected build link location: ${file}`);
      const literal = fs.readlinkSync(path.join(target, file), { encoding: "buffer" });
      built.push([file, "symlink", literal.toString("base64")]);
      assert(
        literal.length > 0 && !path.isAbsolute(literal.toString("utf8")),
        `Expected POSIX relative dependency link: ${file}`,
      );
    } else throw new Error(`Unexpected build entry: ${file}`);
  }
}
try {
  if (fs.existsSync(path.join(target, "dist"))) {
    assert(fs.lstatSync(path.join(target, "dist")).isDirectory(), "Expected a real dist directory");
    visit("dist");
  }
  inventoryComplete = true;
} catch (error) {
  inventoryFailure = { message: String(error.message), code: error.code };
  throw error;
} finally {
  process.stdout.write(
    `${JSON.stringify(
      {
        source: packet.source,
        phase,
        node: process.version,
        sourceHashes: { ...packet.sourceHashes, ...overlay },
        buildFiles: built.filter((row) => row[1] === "file").length,
        buildLinks: built.filter((row) => row[1] === "symlink").length,
        buildHash: hash(JSON.stringify(built)),
        buildInventory: built,
        inventoryComplete,
        inventoryFailure,
      },
      null,
      2,
    )}\n`,
  );
}
